File: src/distance_estimator.py
import math

import cv2

from distance_formula import (
    distance_from_ratio,
    distance_from_width_ratio,
    height_ratio
)


class DistanceEstimator:
    """
    Estimates the player-camera distance from several
    independent measurements and combines them.

    Approaches:
      1. Full body height   - the whole detection box height.
      2. Upper body (head)  - the head detected inside the
                              top of the detection box.
      3. Shoulder width     - the detection box width.

    Each approach returns a distance; the estimator compares
    them and combines the valid ones into a single number.
    """

    # ------------------------------------------------------------
    # Body proportion constants (fractions of body height)
    # ------------------------------------------------------------

    # Average head-to-body ratio (~7.5 heads tall).
    HEAD_RATIO = 1.0 / 7.5

    # Average shoulder width ratio (~0.25 of body height).
    SHOULDER_RATIO = 0.25

    # The detection box is wider than the actual shoulders,
    # because it also covers the arms. Scale it down.
    BOX_TO_SHOULDER_RATIO = 0.8


    def __init__(self, horizontal_fov, player_height=1.75, face_model=None):
        """
        Args:
            horizontal_fov:
                Horizontal field of view of the camera in degrees.

            player_height:
                Real height of the tracked player in meters.

            face_model:
                Path to a YuNet face detection ONNX model.
                When not given, the default bundled model
                (models/yunet_face_detection.onnx) is used.
        """

        self.horizontal_fov = horizontal_fov

        self.player_height = player_height

        self._face_detector = None

        self._face_model = face_model


    def _load_face_detector(self):
        """
        Lazily build the YuNet face detector.
        """

        if self._face_detector is not None:

            return self._face_detector

        if self._face_model is None:

            # Fall back to the bundled model next to the project.

            from pathlib import Path

            default = (
                Path(__file__).resolve().parent.parent
                / "models"
                / "yunet_face_detection.onnx"
            )

            self._face_model = str(default)

        if not cv2.os.path.exists(self._face_model):

            return None

        # Input size is updated per frame in detect_head_height,
        # so any placeholder is fine here.

        self._face_detector = cv2.FaceDetectorYN.create(
            self._face_model,
            "",
            (320, 320)
        )

        return self._face_detector


    def calculate_vertical_fov(self, image_height, image_width):
        """
        Derive the vertical field of view from the
        horizontal field of view and the frame dimensions.
        """

        horizontal_rad = math.radians(
            self.horizontal_fov / 2
        )

        vertical_rad = math.atan(
            math.tan(horizontal_rad) *
            (image_height / image_width)
        )

        return 2 * math.degrees(vertical_rad)


    def is_truncated(
        self,
        detection,
        image_height,
        tolerance=2
    ):
        """
        Check whether the detection box is cut off by the
        top or bottom edge of the frame.
        """

        y1 = detection["y1"]

        y2 = detection["y2"]

        return (
            y1 <= tolerance
            or y2 >= image_height - 1 - tolerance
        )


    # ------------------------------------------------------------
    # Approach 1: full body height
    # ------------------------------------------------------------

    def estimate_from_full_height(
        self,
        detection,
        image_height,
        image_width
    ):
        """
        Distance from the whole detection box height.

        Returns None when the box is truncated.
        """

        if self.is_truncated(
            detection,
            image_height
        ):

            return None

        box_height = (
            detection["y2"] - detection["y1"]
        )

        if box_height <= 0:

            return None

        ratio = height_ratio(
            box_height,
            image_height
        )

        vertical_fov = (
            self.calculate_vertical_fov(
                image_height,
                image_width
            )
        )

        return distance_from_ratio(
            self.player_height,
            vertical_fov,
            ratio
        )


    # ------------------------------------------------------------
    # Approach 2: upper body (head) height
    # ------------------------------------------------------------

    def detect_head_height(
        self,
        frame,
        detection
    ):
        """
        Find the head inside the top half of the detection box
        using a YuNet face detector.

        Returns:
            Head height in pixels, or None when no face is found.
        """

        detector = self._load_face_detector()

        if detector is None:

            return None

        x1 = max(detection["x1"], 0)

        y1 = max(detection["y1"], 0)

        x2 = detection["x2"]

        y2 = detection["y2"]

        # Look for the head in the top half of the box.

        head_bottom = y1 + (y2 - y1) // 2

        if head_bottom <= y1:

            return None

        height, width = frame.shape[:2]

        # FaceDetectorYN input size must match the frame.

        detector.setInputSize((width, height))

        status, faces = detector.detect(frame)

        if status is False or faces is None:

            return None

        best = None

        best_area = 0

        for face in faces:

            fx1, fy1, fw, fh = (
                face[:4].astype(int)
            )

            # Keep only faces inside the head region.

            if (
                fx1 < x1 or fy1 < y1
                or fx1 + fw > x2
                or fy1 + fh > head_bottom
            ):

                continue

            area = fw * fh

            if area > best_area:

                best_area = area

                best = fh

        if best is None or best <= 0:

            return None

        return int(best)


    def estimate_from_head(
        self,
        frame,
        detection,
        image_height,
        image_width
    ):
        """
        Distance from the head height (upper body).

        The head height in the image is detected with a face
        cascade; the real head height is estimated from the
        player's height using the head-to-body ratio.
        """

        head_px = self.detect_head_height(
            frame,
            detection
        )

        if head_px is None or head_px <= 0:

            return None

        real_head = (
            self.player_height * self.HEAD_RATIO
        )

        ratio = height_ratio(
            head_px,
            image_height
        )

        vertical_fov = (
            self.calculate_vertical_fov(
                image_height,
                image_width
            )
        )

        return distance_from_ratio(
            real_head,
            vertical_fov,
            ratio
        )


    # ------------------------------------------------------------
    # Approach 3: shoulder width
    # ------------------------------------------------------------

    def estimate_from_width(
        self,
        detection,
        image_height,
        image_width
    ):
        """
        Distance from the detection box width, scaled to the
        average shoulder width.

        This works even when the box is truncated vertically,
        because the width is still fully visible.
        """

        box_width = (
            detection["x2"] - detection["x1"]
        )

        if box_width <= 0:

            return None

        shoulder_px = (
            box_width * self.BOX_TO_SHOULDER_RATIO
        )

        ratio = (
            shoulder_px / image_width
        )

        real_shoulders = (
            self.player_height * self.SHOULDER_RATIO
        )

        return distance_from_width_ratio(
            real_shoulders,
            self.horizontal_fov,
            ratio
        )


    # ------------------------------------------------------------
    # Compare and combine
    # ------------------------------------------------------------

    def estimate(
        self,
        frame,
        detection,
        image_height,
        image_width
    ):
        """
        Run every available approach, then combine the results.

        Returns:
            (combined_distance, estimates) where estimates is a
            dict of approach name -> distance for comparison.
            combined_distance is None when nothing is usable.
        """

        estimates = {}

        full = self.estimate_from_full_height(
            detection,
            image_height,
            image_width
        )

        if full is not None:

            estimates["full_height"] = full

        head = self.estimate_from_head(
            frame,
            detection,
            image_height,
            image_width
        )

        if head is not None:

            estimates["head"] = head

        width = self.estimate_from_width(
            detection,
            image_height,
            image_width
        )

        if width is not None:

            estimates["shoulder_width"] = width

        if not estimates:

            return None, estimates

        # Combine: median when possible (robust to one bad
        # estimate), otherwise the average of what we have.

        values = list(estimates.values())

        values.sort()

        middle = len(values) // 2

        if len(values) % 2 == 1:

            combined = values[middle]

        elif len(values) == 2:

            combined = (values[0] + values[1]) / 2.0

        else:

            combined = (
                (values[middle - 1] + values[middle]) / 2.0
            )

        return combined, estimates

File: cpp/src/distance_estimator.cpp
#include "distance_estimator.hpp"

#include <algorithm>
#include <cmath>
#include <filesystem>

#include "config.hpp"
#include "distance_formula.hpp"

namespace tp {

namespace {
constexpr double kPi = 3.14159265358979323846;
}  // namespace

DistanceEstimator::DistanceEstimator(double horizontal_fov,
                                     double player_height,
                                     std::string face_model)
    : horizontal_fov_(horizontal_fov),
      player_height_(player_height),
      face_model_(std::move(face_model)) {}

cv::Ptr<cv::FaceDetectorYN> DistanceEstimator::loadFaceDetector() {
    if (!face_detector_.empty()) {
        return face_detector_;
    }

    // Fall back to the bundled model next to the project.
    std::string model = face_model_;
    if (model.empty()) {
        model = (Config::projectDirectory() / "models" /
                 "yunet_face_detection.onnx").string();
    }
    if (!std::filesystem::exists(model)) {
        return cv::Ptr<cv::FaceDetectorYN>();
    }

    // Input size is updated per frame in detectHeadHeight, so any
    // placeholder is fine here.
    try {
        face_detector_ = cv::FaceDetectorYN::create(model, "", cv::Size(320, 320));
    } catch (const cv::Exception&) {
        return cv::Ptr<cv::FaceDetectorYN>();
    }
    return face_detector_;
}

double DistanceEstimator::calculateVerticalFov(int image_height,
                                               int image_width) const {
    const double horizontal_rad = (horizontal_fov_ / 2.0) * (kPi / 180.0);
    const double vertical_rad = std::atan(
        std::tan(horizontal_rad) *
        (static_cast<double>(image_height) / image_width));
    return 2.0 * vertical_rad * (180.0 / kPi);
}

bool DistanceEstimator::isTruncated(const Detection& detection,
                                    int image_height, int tolerance) const {
    return detection.y1 <= tolerance ||
           detection.y2 >= image_height - 1 - tolerance;
}

std::optional<double> DistanceEstimator::estimateFromFullHeight(
    const Detection& detection, int image_height, int image_width) const {
    if (isTruncated(detection, image_height)) {
        return std::nullopt;
    }

    const int box_height = detection.y2 - detection.y1;
    if (box_height <= 0) {
        return std::nullopt;
    }

    const double ratio = heightRatio(box_height, image_height);
    const double vertical_fov = calculateVerticalFov(image_height, image_width);
    return distanceFromRatio(player_height_, vertical_fov, ratio);
}

std::optional<int> DistanceEstimator::detectHeadHeight(
    const cv::Mat& frame, const Detection& detection) {
    const cv::Ptr<cv::FaceDetectorYN> detector = loadFaceDetector();
    if (detector.empty()) {
        return std::nullopt;
    }

    const int x1 = std::max(detection.x1, 0);
    const int y1 = std::max(detection.y1, 0);
    const int x2 = detection.x2;
    const int y2 = detection.y2;

    // Look for the head in the top half of the box.
    const int head_bottom = y1 + (y2 - y1) / 2;
    if (head_bottom <= y1) {
        return std::nullopt;
    }

    // FaceDetectorYN input size must match the frame.
    detector->setInputSize(cv::Size(frame.cols, frame.rows));

    cv::Mat faces;
    detector->detect(frame, faces);
    if (faces.empty()) {
        return std::nullopt;
    }

    int best = -1;
    int best_area = 0;
    for (int i = 0; i < faces.rows; ++i) {
        const int fx1 = static_cast<int>(faces.at<float>(i, 0));
        const int fy1 = static_cast<int>(faces.at<float>(i, 1));
        const int face_width = static_cast<int>(faces.at<float>(i, 2));
        const int face_height = static_cast<int>(faces.at<float>(i, 3));

        // Keep only faces inside the head region.
        if (fx1 < x1 || fy1 < y1 || fx1 + face_width > x2 ||
            fy1 + face_height > head_bottom) {
            continue;
        }

        const int area = face_width * face_height;
        if (area > best_area) {
            best_area = area;
            best = face_height;
        }
    }

    if (best <= 0) {
        return std::nullopt;
    }
    return best;
}

std::optional<double> DistanceEstimator::estimateFromHead(
    const cv::Mat& frame, const Detection& detection,
    int image_height, int image_width) {
    const std::optional<int> head_px = detectHeadHeight(frame, detection);
    if (!head_px.has_value() || *head_px <= 0) {
        return std::nullopt;
    }

    const double real_head = player_height_ * HEAD_RATIO;
    const double ratio = heightRatio(static_cast<double>(*head_px),
                                     static_cast<double>(image_height));
    const double vertical_fov = calculateVerticalFov(image_height, image_width);
    return distanceFromRatio(real_head, vertical_fov, ratio);
}

std::pair<std::optional<double>, std::map<std::string, double>>
DistanceEstimator::estimate(const cv::Mat& frame,
                            const Detection& detection,
                            int image_height, int image_width) {
    std::map<std::string, double> estimates;

    const std::optional<double> full = estimateFromFullHeight(
        detection, image_height, image_width);
    if (full.has_value()) {
        estimates["full_height"] = *full;
        return {full, estimates};
    }

    const std::optional<double> head = estimateFromHead(
        frame, detection, image_height, image_width);
    if (head.has_value()) {
        estimates["head"] = *head;
        return {head, estimates};
    }

    return {std::nullopt, estimates};
}

}  // namespace tp